app/models/railway_stations_route.rb

```diff
@@ -1,4 +1,12 @@
 class RailwayStationsRoute < ApplicationRecord
   belongs_to :railway_station
   belongs_to :route
+  validates :railway_station_id, uniqueness: { scope: :route_id }
+
+  before_save :station_position
+  after_destroy :station_position
+
+  def station_position
+    self.position = route.railway_stations.find_index(railway_station)
```

Comment:
I think better use https://zaiste.net/rails_activerecord_updating_attributes_object_fields/ instead of self.